## Analyzing borrowers’ risk of defaulting

The project is to prepare a report for a bank’s loan division. We’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Our report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

### Step 1. Open the data file and have a look at the general information. 

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/credit_scoring_eng.csv')
print(data.info())
print(data.head(10))
print(data.tail(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None
   children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124

In [2]:
# Use more of the page width
from IPython.core.display import HTML
HTML("<style>.container { width:96% !important; }</style>")

In [3]:
print(data.isnull().sum())
data.duplicated().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


54

In [4]:
##So we only have null values in days_employed and total_income columns. 
##LOOK AT THE VARIABLES

In [5]:
#children
df = data.sort_values('children')['children'].value_counts()
df.sort_index()

-1        47
 0     14149
 1      4818
 2      2055
 3       330
 4        41
 5         9
 20       76
Name: children, dtype: int64

In [6]:
#days_employed, let's see how many negative values are there in our dataset
print(data[data['days_employed']<0]['days_employed'].count())

15906


In [7]:
#dob_years, let's check the how many customers of each age do we have
print(data.sort_values('dob_years')['dob_years'].value_counts().sort_index())
#Notice that we have customers of age 0!

0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64


In [8]:
#education, let's see what type of education do the customers have
data.sort_values('education')['education'].value_counts()
#Notice the duplicates!

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [9]:
#family_status
f = data.sort_values('family_status')['family_status']
print(f.value_counts().sort_index())

civil partnership     4177
divorced              1195
married              12380
unmarried             2813
widow / widower        960
Name: family_status, dtype: int64


In [10]:
#gender
data.sort_values('gender')['gender'].value_counts().sort_index()
#Notice that we have undefined gender!

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [11]:
#income_type
data['income_type'].value_counts()

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
entrepreneur                       2
unemployed                         2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

In [12]:
data['purpose'].value_counts()

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
buying property for renting out             653
housing transactions                        653
transactions with commercial real estate    651
purchase of the house                       647
housing                                     647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
purchase of my own house                    620
building a property                         620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

### Conclusion


We run data.info() to see the overview of our dataset. We see that most of the columns have 21525 data points, except for total_income and days_employed, which have 2174 missing data points. Then we proceed and look at first and last 10 rows to see the values in our dataset. We can immediately see that days_employed has a lot of negative values; however, since a person at worst case hasn't worked at all, the minimum value in the column could possibly be only zero. So we could assume, for example, that our colleagues who filled the data, maybe wanted to use hyphen to distinguish the data, but it was passed as a minus sign. Or, maybe the customers somehow reported misleading data. If we were to use it as an explanatory variable later on, we could "fix" the dataset, by multiplying the negative values by -1. We see that we also have 2174 missing entries in total_income column. Since we need to use total_income in our analysis, the 2174 data points will be ignored, as we can't restore them.

Summary for each individual variable:
- we have 47 negative values reported in the number of children column. Since this amount is not significant we could either ignore the data, or multiply it by -1 and add to a category of people with one child
- Days employed has negative values, and so we will multiply them by -1, to use in analysis
- Dob years, has missing values. We can't assume what people's ages were so due to a small amount we could ignore the data.
- We see several duplicates in education category, so we should genralize the categories.
- We could also note, for the sake of clarity, that family_status_id == 0 refers to married people, family_status_id == 1 refers to people in civic relationships, family_status_id==2 refers to widdowed people, family_status_id==3 refers to divorced people, and family_status_id==4 refers to unmarried people. 
- Purpose column has a lot of duplicates, with different wordings. Again, we will get rid of the duplciates and generalize categories

### Step 2. Data preprocessing

### Processing missing values

In [13]:
print(data['gender'].value_counts())
print(data.loc[data['gender']=='XNA'])
data.set_value(10701, 'gender', 'F')
print(data['gender'].value_counts())

F      14236
M       7288
XNA        1
Name: gender, dtype: int64
       children  days_employed  dob_years     education  education_id  \
10701         0   -2358.600502         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701     32624.825  buy real estate  
F    14237
M     7288
Name: gender, dtype: int64


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  This is separate from the ipykernel package so we can avoid doing imports until


In [14]:
data = data[data['children']>-1]
print(data['children'].value_counts())

0     14149
1      4818
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64


In [15]:
mn = data['total_income'].mean()
data.loc[data['total_income'].isnull(), 'total_income'] = mn

In [16]:
dm = data['days_employed'].mean()
data.loc[data['days_employed'].isnull(), 'days_employed'] = dm

In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21478 entries, 0 to 21524
Data columns (total 12 columns):
children            21478 non-null int64
days_employed       21478 non-null float64
dob_years           21478 non-null int64
education           21478 non-null object
education_id        21478 non-null int64
family_status       21478 non-null object
family_status_id    21478 non-null int64
gender              21478 non-null object
income_type         21478 non-null object
debt                21478 non-null int64
total_income        21478 non-null float64
purpose             21478 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


### Conclusion

We see that we have a person with third unassigned gender XNA, and then we look at the persons data. We see that we have no missing variable there, so for the sake of clarity we could assign this person a female (F) gender, because gender is not the primary focus of our analysis.
We also see that we have some rows where the number of children is reported to be negative, which is not possible, so we delete those rows. 

### Data type replacement

In [18]:
data['days_employed'] = abs(data['days_employed'])    
data['days_employed'] = data['days_employed'].astype(int)
len(data[data['days_employed']<0])

0

### Conclusion

The primary goal was to replace the negative values in days_employed column, to improve our analysis. As we see at the end there no negative values left in the column, as we took their absolute value. Also we replaced the data type from float to int.

### Processing duplicates

In [19]:
#We have duplicates in education column
data['education'] = data['education'].str.lower()
print(data['education'].value_counts())

secondary education    15196
bachelor's degree       5251
some college             743
primary education        282
graduate degree            6
Name: education, dtype: int64


### Conclusion

Due to the difference in the letter sizes we had several education categories. We decreased them substantially, and now have separate categories.

### Lemmatization

In [34]:
from collections import Counter
import nltk
from nltk.stem import WordNetLemmatizer
lem = WordNetLemmatizer()
#Originally, I wanted to lemmatize purposes, take out the most coomon ones and categorize according to them.I did not succeed, so the code  her ewas utilized from that posted by others in Slack channel, so 
w_words = ['wedding']
r_words = ['real estate', 'transaction', 'transactions', 'property']
h_words = ['house', 'housing']
c_words = ['car', 'cars']
s_words = ['education', 'educated']
def purpose_grouping(purpose):       
    for w in w_words:
        if w in purpose:
                return 'wedding'
    for r in r_words:
        if r in purpose:
            return 'real estate'
    for h in h_words:
        if h in purpose:
            return 'house'
    for c in c_words:
        if c in purpose:
            return 'car'
    for s in s_words:
        if s in purpose:
            return 'education'
data['purpose_type'] = data['purpose'].apply(purpose_grouping)

data['purpose_type'].value_counts()

real estate    7655
car            4304
education      3516
house          3161
wedding        2346
Name: purpose_type, dtype: int64

### Conclusion

We boiled down the purposes to 5 main categories, which makes analysis much more simpler. 

### Categorizing Data

In [36]:
def child_categories(n):
    if n==0:
        return 'childless'
    elif n<3:
        return 'small family'
    else:
        return 'big family'

data['child_type'] = data['children'].apply(child_categories)
#Let's see what categorie we got
data['child_type'].value_counts()

childless       14149
small family     6873
big family        456
Name: child_type, dtype: int64

In [37]:
def family_categories(string):
    if string == 'married':
        return 'married'
    else:
        return 'unmarried'
    
data['family_status_type'] = data['family_status'].apply(family_categories)
#Let's see how the distribution looks like
data['family_status_type'].value_counts()

married      12351
unmarried     9127
Name: family_status_type, dtype: int64

In [45]:
from scipy.stats import iqr
import numpy as np
#the lowest 25% of the distribution is smaller than p25
p25 = np.percentile(data['total_income'],25)
#the highest 25% of the distribution is above p75
p75 = np.percentile(data['total_income'],75)
def income_categories(n):
    if n<=p25:
        return 'low income'
    elif n<=p75:
        return 'average income'
    else:
        return 'high income'
    
data['total_income_type'] = data['total_income'].apply(income_categories) 
data.head()

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose purpose_type    child_type family_status_type  \
0    purchase of the house        house  small family            married   
1             car purchase          car  small family            married   
2    purchase of the house        house     childless            married   
3  supplementary education    education    big family            married   
4        to have a wedding      wedding     childless          unmarried   

  total_income_type  
0       high income  
1    average income  
2    average income  
3       high income  
4    average income

### Conclusion

We categorized the data to see :
-Is there a relationship between the number of children, marital status, total income and loan repayment rate

### Step 3. Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [46]:
pd.pivot_table(data, index='child_type', values='debt')

debt
child_type            
big family    0.085526
childless     0.075129
small family  0.092827

### Conclusion

There is no direct relationship between the number of children and the loan repayment rate. Families with no children, and big families with many children have a difference of 1% on their debt which is not significant. In addition, families which have less than 3 children tend to have higher debts, than others. Therefore, we cannot state that increase/decrease in the number of children hav a direct effect

- Is there a relation between marital status and repaying a loan on time?

In [47]:
pd.pivot_table(data, index='family_status_type', values='debt')

debt
family_status_type          
married             0.075298
unmarried           0.088748

### Conclusion

Unmarried people tend to have higher debt rate, i.e lower loan repayment rate

- Is there a relation between income level and repaying a loan on time?

In [48]:
pd.pivot_table(data, index='total_income_type', values='debt')

debt
total_income_type          
average income     0.086422
high income        0.071695
low income         0.079516

### Conclusion

People with low and high incomes tend to be more accurate with their repayments (i.e repay on time), whereas, people with average income tend to have higher debt rates

- How do different loan purposes affect on-time repayment of the loan?

In [49]:
pd.pivot_table(data, index='purpose_type', values='debt')

debt
purpose_type          
car           0.093401
education     0.093003
house         0.065802
real estate   0.074984
wedding       0.079284

### Conclusion

People mainly default on car and education purposes. Housing has the lowest debt rate. Real estate and wedding are in the middle, and have 7.49% and 7.92% debt rates respectively.

### Step 4. General conclusion

The bank will have a lower risk if it gives a loan to people with no children or a lot of children, who are married, who have low or high income, and are intending to buy a house, or possibly real estate or are willing to have a wedding.